[JSC] B3 Values shouldn't recompute their adjacency list offset when children are accessed

ddegazio opened the pull request:
#### 6bca30fcc772588914e2db2da8f26c69f1c7eecd
<pre>
[JSC] B3 Values shouldn&apos;t recompute their adjacency list offset when children are accessed
<a href="https://bugs.webkit.org/show_bug.cgi?id=277381">https://bugs.webkit.org/show_bug.cgi?id=277381</a>
<a href="https://rdar.apple.com/132848864">rdar://132848864</a>

Reviewed by Yusuke Suzuki.

Avoids calling adjacencyListOffset() every time we need to access the adjacency list
of a B3::Value. Currently, because we don&apos;t have the information about what derived
class we are, this function means we switch on the B3::Value&apos;s opcode every time we
access any child of a Value. With this patch, we instead compute this offset once on
node construction, and cache it as a 16-bit integer field on the base Value, in what
was previously two bytes of unused padding.

* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
(JSC::B3::Value::computeAdjacencyListOffset const):
(JSC::B3::Value::adjacencyListOffset const): Deleted.

Canonical link: <a href="https://commits.webkit.org/281663@main">https://commits.webkit.org/281663@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d17e319d47613bdc6753fd668d84cebdfca7d04e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60475 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13042 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64398 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11012 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47506 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11243 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48934 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7651 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62506 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37103 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52381 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29767 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33800 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9613 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9927 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/53573 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55682 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9908 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66129 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/59720 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4413 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9738 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56300 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4432 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52359 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56462 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13498 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3665 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/81477 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35638 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14151 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36720 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37811 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36465 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/b3/B3Value.h:
```diff
@@ -611,6 +611,10 @@ class JS_EXPORT_PRIVATE Value {
     template<typename... Arguments>
     void buildAdjacencyList(NumChildren numChildren, Arguments... arguments)
     {
+        size_t offset = computeAdjacencyListOffset();
+        RELEASE_ASSERT(offset == static_cast<uint16_t>(offset));
+        m_adjacencyListOffset = offset;
+
```
Can you check whether this function is always called for all B3::Value creation?
Also, please check that this is correctly setup even when we refresh the current B3::Value in-place. (We use `new (this) Value` when we replace the node. See Value::replaceWith)

Constellation replied:
Probably it is done correctly, but let's double-check :)

ddegazio replied:
Because we compute the offset in `buildAdjacencyList`, anywhere we are initializing the adjacency list should correctly set the offset. Specifically, we do this in every constructor of `Value` except the copy constructor. In the copy constructor's case though, it's a protected member of `Value` and only called from `Value::allocate` (which is what we call when cloning values). For `replaceWith`, we call a normal `Value` constructor, so we `buildAdjacencyList` and set the offset. So, we always set this field before we access the adjacency list of a value.

ddegazio committed:
[JSC] B3 Values shouldn't recompute their adjacency list offset when children are accessed
https://bugs.webkit.org/show_bug.cgi?id=277381
rdar://132848864

Reviewed by Yusuke Suzuki.

Avoids calling adjacencyListOffset() every time we need to access the adjacency list
of a B3::Value. Currently, because we don't have the information about what derived
class we are, this function means we switch on the B3::Value's opcode every time we
access any child of a Value. With this patch, we instead compute this offset once on
node construction, and cache it as a 16-bit integer field on the base Value, in what
was previously two bytes of unused padding.

* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
(JSC::B3::Value::computeAdjacencyListOffset const):
(JSC::B3::Value::adjacencyListOffset const): Deleted.

Canonical link: https://commits.webkit.org/281663@main
